resolver/upstream_resolver.go

```diff
@@ -206,11 +244,25 @@ func (r *UpstreamResolver) Resolve(request *model.Request) (response *model.Resp
 		}),
 		retry.OnRetry(func(n uint, err error) {
 			logger.WithField("attempt", fmt.Sprintf("%d/%d", n+1, retryAttempts)).
+				WithField("failed_ip", ip.String()).
 				Debugf("Temporary network error / Timeout occurred, retrying...")
+
+			ip = r.nextIP()
 		}))
 	if err != nil {
 		return nil, err
 	}
 
-	return &model.Response{Res: resp, Reason: fmt.Sprintf("RESOLVED (%s)", r.upstreamURL)}, nil
+	return &model.Response{Res: resp, Reason: fmt.Sprintf("RESOLVED (%s)", upstreamURL)}, nil
+}
+
+func (r *UpstreamResolver) nextIP() net.IP {
```

Comment:
I have a first implementation of handling everything in `Boostrap` locally but it still needs improvement.
`UpstreamResolver`s will have a reference to `Bootstrap` and use it to get an IP at `Resolve`.
The one weird part is that the function to get an IP takes the resolver and not just a hostname because `Bootstrap` needs to be able to check if the resolver is part of its internal resolver, to avoid infinite recursion.
For now, in that case we just use the hardcoded IPs and never refresh them.

I ported over the `nextIP` code, but to do it properly there, it requires much more complexity so I'm not sure it's worth it: indexes must be stored in a thread-safe map (which means new dependency), and is more complex to update than just a compare-and-swap.
I didn't take the time to do it properly, it's one of the things I need to clean, so I might just remove that.

Reply:
I pushed the updated version.

I created a wrapper type to have the logic of `nextIP` in boostrap.go, but there's no getting around calling it in `UpstreamResolver` if we want the code to be simple.
Also it doesn't work as well as the first version since the IP array and index are reset on each `Resolve`, so the "good" ip is never actually saved. We just don't reuse a failing IP when others are available.
If you think it's not worth it I'll just completely remove this.

Besides that, I think everything is looking good now. Missing docs and tests though.

Reply:
ok, thanks. I'll take a look this week 👍 🤝